****---
title: "StatBI/E Week 1"
author: ''
output:
  html_notebook:
    code_folding: none
    toc: true
    toc_float:
      collapsed: false
    df_print: paged
    theme: sandstone
    highlight: kate
  html_document:
    toc: true
    df_print: paged
  pdf_document:
    toc: true
---
<style>
.exercise{background-color:#c9e2a5;padding:1ex 1ex .1ex 1ex;}
.answer{background-color:#f3f3a5;padding:1ex 1ex .1ex 1ex;}
</style>
::: exercise :::
[This is how exercises are highlighted]
:::
::: answer  :::
[This is how your answers are highlighted]
:::




## Exercise 1  A case study of DNA sequence

Let $\Omega$
be the sample space of a single DNA nucleotide type,
\[ \Omega = \{ A, T, C, G \}. \]

Suppose that all four  nucleotide types have equal probabilities in the genome and 
all locations in the DNA are independent. 

A DNA motif is a specific sequence of nucleotides, for example AATG or CGGCC. The number of
nucleotides in the motif is called the length of the motif.


### Exercise 1.1 

::: exercise :::
Describe the sample space of DNA sequences of length 2.
:::

::: answer :::

:::

### Exercise 1.2 

::: exercise :::
Consider a DNA sequence of length 5. Describe the event of observing the 
motif ACG inside the DNA sequence.
:::

::: answer :::

:::

::: exercise :::
What is the probability of this event?
:::

::: answer :::

:::



### Exercise 1.3 

::: exercise :::
Suppose we observe one nucleotide in the genome at a time. We keep 
observing until we find the nucleotide G.  Describe the sample space of the 
observed DNA sequence. 
What is the probability that the DNA sequence observed before observing G has length 
10?
:::

::: answer :::

:::

### Exercise 1.4 

::: exercise :::
Now we drop the independence assumption. Suppose that the 
probability of observing A,T,C or G given the previous nucleotide being C, is
respectively 0.2, 0.2, 0.5, 0.1. In general in different locations of the DNA
sequence, this conditional probabilities will be different and this fact is
important for sequence detection.

1. Suppose we observe C at a given location, what is the probability of
observing the motif CG immediately after?
:::

::: answer :::

:::

::: exercise :::
2. Suppose now in a different region of the DNA sequence, the 
conditional probabilities of A, T, C, G given C are 0.1, 0.2, 0.3, 0.4.
In that DNA sequence region,
what is the probability of observing CG given
that we observed C in the previous location?
:::

::: answer :::

:::



## Exercise 2 Some simulations in R

We now conduct some simulations in R for the DNA case study.

### Exercise 2.1 

::: exercise :::
Run a simulation of Exercise 1.2 and check the probability of observing the motif ACG 
in a sequence of length 5. Idea: generate a sequence of length 5 and check if it contains
the motif ACG. Repeat the above sampling and checking procedure 1000, 10000, or 100000 times. 
For each case obtain the proportion of sequences containing the motif ACG. Use 
these proportions as an approximation of the probability of observing the motif ACG. 
:::

::: answer :::

:::****


### Exercise 2.2

::: exercise :::
Run an R simulation of Exercise 1.3 and calculate the probability of observing 
10 nucleotides before G, using 10000 repetitions.
:::

::: answer :::

:::


### Exercise 2.3 

::: exercise :::
Repeat Exercise 2.2 for different lengths of the sequence $\{0,1,2,\ldots, 20\}$.
Show the result in a scatter plot, with the sequence lengths $\{0,1,\ldots, 20\}$ on the x-axis and the approximated probabilities on the y-axis. Now also add the true probabilities (as derived in Exercise 1.3) to the plot.
:::

::: answer :::

:::
